src/Chapter3.hs

```diff
@@ -482,6 +496,38 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
 
+data City =
+  City {
+    castle:: Maybe Castle,
+    education:: Education,
+    cityHouses:: [House]
+  }
+
+data Castle = Castle { castleName:: String, hasWall:: Bool }
+
+data Education = Church | Library
+
+data House = House { housePeople:: Int }
+
+buildCastle:: String -> City -> City
+buildCastle newName oldCity = case castle oldCity of
+  Nothing ->  oldCity { castle = Just $ Castle newName False }
+  Just (Castle _ oldWall) -> oldCity { castle = Just $ Castle newName oldWall }
+
+buildHouse:: Int -> City -> City
+buildHouse people oldCity = oldCity { cityHouses = House people : cityHouses oldCity }
```

Comment:
Interesting approach. I see that you have a separate data type for `House` but in this function you pass `Int`. This means that a caller of this function can pass any integer number, not only between 1 and 4.

You can improve the type-safety of your solution by using the approach called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂

Reply:
> Generally, in here, you could restrict that on type level, by using something else than Int 

I didn't find reference to literal types anywhere, was trying something like

data People = 1 | 2 | 3 | 4

or do you mean named constructors with some function to translate to Int value, perhaps Enum

Reply:
@gagandeepkalra You can implent it something like this:

```haskell
data People = One | Two | Three | Four
```

or like this

```haskell
data People = P1 | P2 | P3 | P4
```

You can't use `|` between values, you need to have tags for name constructors.